### PR TITLE
feat: not auto href matching when selecting for the toolbar

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

see #10838 
stacked on top of https://github.com/PostHog/posthog/pull/13864

When matching on elements the toolbar always includes any href selector if present. This means that elements whose link changes cannot be matched.

E.g. the third item in a list will be different for different views of the page or the profile button on a web page will have a unique id in the href

## Changes

* by default not include href in selectors
* let user turn on href matching if they want to
* fixes tooltips for the `HeatmapStats` window 
* fly-by de-antsing of the `HeatmapStats window

matching-by-link

|with ant|without ant|
|-|-|
|Screenshot 2023-01-22 at 14 14 13|Screenshot 2023-01-22 at 14 35 49|

## How did you test this code?

running it locally and 👀

### CODE REVIEW

File: frontend/src/toolbar/stats/HeatmapStat.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.HeatmapStats__elements-list:hover {
+    background-color: var(--primary-highlight);
+}
```
Comment (mariusandra) [inline frontend/src/toolbar/stats/HeatmapStat.scss:3]: This is totally fine, but if I were to create more work, I'd suggest adding some SCSS magic to autogenerate a bunch of `hover:bg-primary` [tailwindesque](https://tailwindcss.com/docs/background-color#hover-focus-and-other-states) color classes in utilities.scss 🙃

File: frontend/src/toolbar/stats/HeatmapStats.tsx [hunk 2]
```diff
@@ -9,15 +8,66 @@ import { LemonInput } from 'lib/components/LemonInput/LemonInput'
...
+ * LemonButton can accept a `disabledReason` prop, but it attaches its tooltip to the body
+ *
+ * This component wraps the LemonButton in a div, and attaches the tooltip to that div
+ * and passes the correct stacking context so that the tooltip is visible
+ * */
+function ButtonWithTooltipInStackingContext(props: {
+    canLoadMoreElementStats: any
+    onClick: () => any
+    popupContainer: () => any
+}): JSX.Element {
```
Comment (mariusandra) [inline frontend/src/toolbar/stats/HeatmapStats.tsx:28]: A lot of the rest of the code uses this pattern to solve the same problem: `getPopupContainer={getShadowRootPopupContainer}`.

Can we add `getPopupContainer` as something we pass to `LemonButton`, and just set it accordingly on the tooltip?

We could also use a React context to pass this around 🤔

File: frontend/src/toolbar/elements/heatmapLogic.ts [hunk 3]
```diff
@@ -149,14 +151,19 @@ export const heatmapLogic = kea<heatmapLogicType>([
...
+                selectors.matchLinksByHref,
+            ],
+            (elementStats, dataAttributes, pageElements, matchLinksByHref) => {
                 const elements: CountedHTMLElement[] = []
                 elementStats?.results.forEach((event) => {
                     let combinedSelector: string
                     let lastSelector: string | undefined
                     for (let i = 0; i < event.elements.length; i++) {
-                        const selector = elementToSelector(event.elements[i], dataAttributes) || '*'
+                        const selector = elementToSelector(event.elements[i], dataAttributes, matchLinksByHref) || '*'
```
Comment (mariusandra) [inline frontend/src/toolbar/elements/heatmapLogic.ts:166]: What about `

```suggestion
                        const element = event.elements[i]
                        const selector = elementToSelector(matchLinksByHref ? element : { ...element, href: undefined }, dataAttributes) || '*'
```

That way the `elementToSelector` function doesn't need to know about the changes you want to make to the element.

File: frontend/src/toolbar/button/ButtonWindow.tsx [hunk 4]
```diff
@@ -22,6 +25,14 @@ export function ButtonWindow({
     tagComponent,
     children,
 }: ButtonWindowProps): JSX.Element {
+    const windowRef = useRef<HTMLDivElement | null>(null)
+    const { storeButtonWindowRef } = useActions(toolbarButtonLogic)
+    useEffect(() => {
+        // store a ref to the window here
+        // it can be used in children to for e.g. ensure tooltips are visible
+        storeButtonWindowRef(windowRef)
+    }, [storeButtonWindowRef, windowRef.current, visible])
```
Comment (mariusandra) [inline frontend/src/toolbar/button/ButtonWindow.tsx:34]: Why not `getShadowRootPopupContainer` ?

File: frontend/src/toolbar/stats/HeatmapStats.tsx [hunk 5]
```diff
@@ -41,58 +91,53 @@ export function HeatmapStats(): JSX.Element {
...
+                    <Tooltip
+                        title={
+                            'Matching links by their target URL can exclude clicks from the heatmap if the URL is too unique.'
+                        }
+                        getPopupContainer={() => buttonWindowRef?.current ?? document.body}
+                    >
+                        <div>
+                            <LemonSwitch
+                                checked={matchLinksByHref}
+                                label={'Match links by their target URL'}
```
Comment (mariusandra) [inline frontend/src/toolbar/stats/HeatmapStats.tsx:110]: When I read this switch label (on the screencast) "match links with urls? 🧐", I got confused. The tooltip also didn't help. Unfortunately I'm also a developer who struggles with writing clear interface text. Fortunately we have @Twixes 😅

**Review by mariusandra — CHANGES_REQUESTED** [review]
I think you could do the same in 1/3 of the lines 😅

File: frontend/src/toolbar/stats/HeatmapStat.scss (hunk 1 above)
Comment (pauldambra) [reply frontend/src/toolbar/stats/HeatmapStat.scss:3]: OMG I opened the file, pressed enter and github co-pilot suggested

```
@each $name, $hex in $colors {
    .hover\\:text-#{$name}:hover {
        color: $hex;
    }
    .hover\\:bg-#{$name}:hover {
        background-color: $hex;
    }
    .hover\\:border-#{$name}:hover {
        border-color: $hex;
    }
}
```

How does it know?!

https://github.com/PostHog/posthog/pull/13870

Comment (mariusandra) [reply frontend/src/toolbar/stats/HeatmapStat.scss:3]: The time AI saves is offset by the time developers are amazed at how much time AI saves

File: frontend/src/toolbar/button/ButtonWindow.tsx (hunk 4 above)
Comment (pauldambra) [reply frontend/src/toolbar/button/ButtonWindow.tsx:34]: I didn't know it existed 🤣 👀

File: frontend/src/toolbar/stats/HeatmapStat.scss (hunk 1 above)
Comment (pauldambra) [reply frontend/src/toolbar/stats/HeatmapStat.scss:3]: me: *saves 4 seconds using co-pilot
also me: shares screenshot in slack, mastodon, twitter, and github UI taking 7 minutes

**Review by mariusandra — APPROVED** [review]

**Pull request merged**